src/librustc/ty/layout.rs

```diff
@@ -1694,10 +1686,11 @@ impl<'a, 'tcx, C> TyLayoutMethods<'tcx, C> for Ty<'tcx>
                     } else {
                         tcx.mk_mut_ref(tcx.types.re_static, nil)
                     };
-                    return cx.layout_of(ptr_ty).map_same(|mut ptr_layout| {
+                    let ptr_layout: C::TyLayout = cx.layout_of(ptr_ty);
```

Comment:
@oli-obk I get this error here:
```
error[E0308]: mismatched types==============================>        ] 118/137: rustc                                                                                       
    --> src/librustc/ty/layout.rs:1692:32
     |
1656 |     fn field(this: TyLayout<'tcx>, cx: &C, i: usize) -> C::TyLayout {
     |                                                         ----------- expected `<C as rustc_target::abi::LayoutOf>::TyLayout` because of return type
...
1692 |                         return ptr_layout;
     |                                ^^^^^^^^^^ expected associated type, found struct `rustc_target::abi::TyLayout`
     |
     = note: expected type `<C as rustc_target::abi::LayoutOf>::TyLayout`
                found type `rustc_target::abi::TyLayout<'_, &ty::TyS<'_>>`
```

Why is `to_result` changing the type here?

Reply:
The error is sadly not telling you what the actual issue is. What you want here is to convert the result back to a `MaybeResult` after mapping it

```rust
return MaybeResult::from(cx.layout_of(ptr_ty).to_result().map(|mut ptr_layout| {
    ptr_layout.ty = this.ty;
    ptr_layout 
}));
```

Reply:
@oli-obk Should this be raised as an issue? :o

Reply:
I think we used to have some more aggressive suggestions around that. If you can create a small example, then a diagnostics issue would be great, yes.